### PR TITLE
[expo-av] Add pitch algorithm to set rate

### LINKED ISSUE #3197: Audio.Sound timestretching has heavily degraded quality on iOS (but not Android)
Reproducible demo: https://snack.expo.io/@jiveworld/jiveworld-expo-test

We are slowing down the playback (timestretching) an mp3 playback, using 

`Audio.Sound.setRateAsync(...)`

The sound quality is excellent when running on Android, but is severely degraded with "echo" artifacts when running on iOS.

### Environment
Target is iOS and Android - but problem only on iOS.

❯ expo diagnostics                                                                                                                                                                                              
  Expo CLI 2.6.14 environment info:
    System:
      OS: macOS 10.14.2
      Shell: 5.5.1 - /usr/local/bin/zsh
    Binaries:
      Node: 10.14.1 - /usr/local/bin/node
      Yarn: 1.12.3 - /usr/local/bin/yarn
      npm: 6.5.0 - /usr/local/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    IDEs:
      Android Studio: 3.2 AI-181.5540.7.32.5056338
      Xcode: 9.4.1/9F2000 - /usr/bin/xcodebuild
    npmPackages:
      expo: ^31.0.2 => 31.0.5
      react: 16.5.0 => 16.5.0
      react-native: https://github.com/expo/react-native/archive/sdk-31.0.0.tar.gz => 0.57.1
      react-navigation: ^3.0.8 => 3.0.8
    npmGlobalPackages:
      expo-cli: 2.6.14

### Steps to Reproduce

1. After load, select a track from the list at the top
2. Click `Play`
3. Slide the `Time stretching` slider at the bottom to, for example, 0.7

### Expected Behavior

The playback should be slower without any change in pitch, and without any very noticeable audio artifacts.

### Actual Behavior

On iOS handsets there are very strong artifacts that cause a reverb type effect with heavy early reflections.

On Android there the playback is smooth.

Note also that when running iOS (not Android) it sometimes (not always) get's stuck on the initial loading screen as if it isn't connecting to the server to download the audio. Please retry if that happens. 

### Reproducible Demo

https://snack.expo.io/@jiveworld/jiveworld-expo-test

### PR DESCRIPTION
**Pull request by enterteg**
# Why
Closes https://github.com/expo/expo/issues/3197 - fixed issue with low quality sound when using `rate < 1` and `shouldCorrectPitch=true`.

# How
Added `pitchCorrectionQuality` param to `setRateAsync` method on `Audio.Sound` object which defines which algorithm for time pitch as written in docs https://developer.apple.com/documentation/avfoundation/audio_track_engineering/audio_settings_and_formats/time_pitch_algorithm_settings?language=objc - possible values: `Audio.PitchCorrectionQuality.([Low|Medium|High)`

# Test Plan
Test via NCL, since bug only occurs on voice audio we need to update url in line `AudioScreen.js` to voice mp3 and change default algorithm in `AudioPlayer.js` in line `57` in method `setRateAsync` to `pitchCorrectionQuality = Audio.PitchCorrectionQuality.High` and check if sound has good quality.

### CODE REVIEW

File: packages/expo-av/src/AV.ts [hunk 1]
```diff
@@ -10,6 +10,12 @@ import { Platform } from 'expo-core';
 //  API to select stream type on Android
 //  subtitles API
 
+export enum PitchCorrectionQuality {
+  Low = ExponentAV.Qualities && ExponentAV.Qualities.Low,
```
Comment (tsapeta) [inline packages/expo-av/src/AV.ts]: actually it should be 
```suggestion
  Low = ExponentAV && ExponentAV.Qualities && ExponentAV.Qualities.Low,
```
that's why I dislike this native-constants-to-enum way 😅

File: packages/expo-av/ios/EXAV/EXAVPlayerData.m [hunk 2]
```diff
@@ -156,7 +159,7 @@ - (void)_finishLoadingNewPlayer
           strongSelfInner.currentPosition = strongSelfInner.player.currentTime;
 
           if (strongSelfInner.shouldCorrectPitch) {
-            strongSelfInner.player.currentItem.audioTimePitchAlgorithm = AVAudioTimePitchAlgorithmLowQualityZeroLatency;
+            strongSelfInner.player.currentItem.audioTimePitchAlgorithm = strongSelfInner.pitchCorrectionQuality;
```
Comment (tsapeta) [inline packages/expo-av/ios/EXAV/EXAVPlayerData.m]: probably we can get rid of this if-else and use `AVAudioTimePitchAlgorithmVarispeed` as a default value of `strongSelfInner.pitchCorrectionQuality`

File: packages/expo-av/ios/EXAV/EXAVPlayerData.m [hunk 3]
```diff
@@ -261,6 +264,12 @@ - (void)setStatus:(NSDictionary *)parameters
     NSNumber *rate = parameters[EXAVPlayerDataStatusRateKeyPath];
     _rate = rate;
   }
+  
+  if ([parameters objectForKey:EXAVPlayerDataStatusPitchCorrectionQualityKeyPath] != nil) {
```
Comment (tsapeta) [inline packages/expo-av/ios/EXAV/EXAVPlayerData.m]:
```suggestion
  if (parameters[EXAVPlayerDataStatusPitchCorrectionQualityKeyPath] != nil) {
```

File: packages/expo-av/src/AV.ts [hunk 4]
```diff
@@ -272,8 +283,16 @@ export const PlaybackMixin = {
     });
   },
 
-  async setRateAsync(rate: number, shouldCorrectPitch: boolean): Promise<PlaybackStatus> {
-    return ((this as any) as AV).setStatusAsync({ rate, shouldCorrectPitch });
+  async setRateAsync(
+    rate: number,
+    shouldCorrectPitch: boolean,
```
Comment (tsapeta) [inline packages/expo-av/src/AV.ts]: I guess it should be `false` by default 🤔 
```suggestion
    shouldCorrectPitch: boolean = false,
```

File: apps/test-suite/tests/Audio.js [hunk 5]
```diff
@@ -574,6 +575,7 @@ export function test(t) {
           const status = await soundObject.setRateAsync(rate, shouldCorrectPitch);
```
Comment (tsapeta) [inline apps/test-suite/tests/Audio.js]:
```suggestion
          const status = await soundObject.setRateAsync(rate, shouldCorrectPitch, pitchCorrectionQuality);
```

File: docs/pages/versions/unversioned/sdk/av.md [hunk 6]
```diff
@@ -175,15 +175,23 @@ The following convenience methods built on top of `setStatusAsync()` are also pr
...
 
-    -   **value (_number_)** -- The desired playback rate of the media. This value must be between `0.0` and `32.0`. Only available on Android API version 23 and later and iOS.
+      - **value (_number_)** -- The desired playback rate of the media. This value must be between `0.0` and `32.0`. Only available on Android API version 23 and later and iOS.
 
-    -   **shouldCorrectPitch (_boolean_)** -- A boolean describing if we should correct the pitch for a changed rate. If set to `true`, the pitch of the audio will be corrected (so a rate different than `1.0` will timestretch the audio).
+      - **shouldCorrectPitch (_boolean_)** -- A boolean describing if we should correct the pitch for a changed rate. If set to `true`, the pitch of the audio will be corrected (so a rate different than `1.0` will timestretch the audio).
+
+      - **pitchCorrectionQuality (_Audio.PitchCorrectionQuality_)** -- iOS time pitch algorithm setting, defaults to `Audio.PitchCorrectionQuality.Low`. Available values:
+
+        `Low` - equivalent to `AVAudioTimePitchAlgorithmLowQualityZeroLatency`
```
Comment (tsapeta) [inline docs/pages/versions/unversioned/sdk/av.md]: Can we link to Apple docs? https://developer.apple.com/documentation/avfoundation/avaudiotimepitchalgorithmlowqualityzerolatency

**Review by tsapeta — CHANGES_REQUESTED** [review]

File: packages/expo-av/ios/EXAV/EXAVPlayerData.m [hunk 7]
```diff
@@ -261,6 +264,12 @@ - (void)setStatus:(NSDictionary *)parameters
...
+    NSString *pitchCorrectionQuality = parameters[EXAVPlayerDataStatusPitchCorrectionQualityKeyPath];
```
Comment (sjchmiela) [inline packages/expo-av/ios/EXAV/EXAVPlayerData.m]: Do we need this temporary variable?

**Review by sjchmiela — APPROVED** [review]
Just some minor small comment, thanks a lot for your contribution!

File: packages/expo-av/ios/EXAV/EXAVPlayerData.m (hunk 3 above)
Comment (enterteg) [reply packages/expo-av/ios/EXAV/EXAVPlayerData.m]: Wanted to be consistent with other similar code in this file. Should I change this one line?

Comment (tsapeta) [reply packages/expo-av/ios/EXAV/EXAVPlayerData.m]: yes, please 🙂 we want to make things more consistent across all classes and not just this one

**Review by sjchmiela — APPROVED** [review]

**Pull request merged**